### PR TITLE
Traces: APM table

### LINKED ISSUE #50258: APM table
https://docs.google.com/document/d/1HYPrJ49ab16iXfreaJX4DrMBm4ynkPd_1HAZBwrhmY0

### PR DESCRIPTION
**Pull request by joey-grafana**
The APM table adds application performance management functionality above the traces view service graph, and is based on discussions/improvements with the Tempo team.

This PR also includes a fix to allow the service graph (and so also the APM table) to be viewed in a dashboard.

Screenshot 2022-05-28 at 10 31 32

**Special notes for your reviewer**:

1. You can use [this Gist](https://gist.github.com/joey-grafana/b29163a2c432cb63ca54e562169dfc23) to setup your dev env to run the table
2. Enable tempoServiceGraph & tempoApmTable in your custom.ini

### CODE REVIEW

**Comment by kvrhdn** [comment]
Hi 👋🏻 
This is really neat and works well! A few bits of feedback:

- If you click on a number to open the query in a side panel, is it possible to use a slightly different query in the side panel?
  For instance, if you click on the rate, it queries for:
  ```
  topk(5, sum(rate(traces_spanmetrics_calls_total{span_name="HTTP Client"}[$__range] @ end())) by (span_name))
  ```
  Could you instead use a non-instant query, e.g.:
  ```
  topk(5, sum(rate(traces_spanmetrics_calls_total{span_name="HTTP Client"}[$__rate_interval])) by (span_name))
  ```
  Changes: use `$__rate_interval`, drop `@ end()`
  The first query shows a completely flat result, which can be misleading. Same for the error rate and the duration column.
- The links column doesn't seem to work, it opens Tempo search but doesn't fill in anything
- The unit of the duration column is incorrect I think: it shows values like `0.121 ms`, this should be `121 ms` or `0.121 s` I think. The underlying metric is in seconds.

File: public/app/plugins/datasource/tempo/datasource.test.ts [hunk 1]
```diff
@@ -391,6 +402,146 @@ describe('Tempo data source', () => {
     const lokiDS4 = ds4.getLokiSearchDS();
     expect(lokiDS4).toBe(undefined);
   });
+
+  it('should build expr correctly', () => {
```
Comment (connorlindsey) [inline public/app/plugins/datasource/tempo/datasource.test.ts:406]: Nit: Maybe add APM table tests to a new describe since this is a large test file

**Review by connorlindsey — COMMENTED** [review]
I wonder if we should put this behind a separate feature flag or give the option to hide it in the datasource config. I think it's different enough from the current behavior to warrant that. 

Would also be nice to add some documentation if it's not behind a separate feature flag.

**Comment by joey-grafana** [comment]
> I wonder if we should put this behind a separate feature flag or give the option to hide it in the datasource config. I think it's different enough from the current behavior to warrant that.
> 
> Would also be nice to add some documentation if it's not behind a separate feature flag.

I could add a new flag for it! :) 

> I wonder if we should put this behind a separate feature flag or give the option to hide it in the datasource config. I think it's different enough from the current behavior to warrant that.
> 
> Would also be nice to add some documentation if it's not behind a separate feature flag.

Two good points.

I've added the APM table behind a feature flag.

I've also added docs. The image for the docs is [here](https://github.com/grafana/website/pull/8944).

File: public/app/plugins/datasource/tempo/datasource.ts [hunk 2]
```diff
@@ -399,8 +409,13 @@ function queryServiceMapPrometheus(request: DataQueryRequest<PromQuery>, datasou
   );
 }
 
-function serviceMapQuery(request: DataQueryRequest<TempoQuery>, datasourceUid: string) {
-  return queryServiceMapPrometheus(makePromServiceMapRequest(request), datasourceUid).pipe(
+function firstServiceMapQuery(request: DataQueryRequest<TempoQuery>, datasourceUid: string) {
+  const serviceMapRequest = makePromServiceMapRequest(request);
+  serviceMapRequest.targets = makeApmRequest([buildExpr(rateMetric, '', request)]).concat(
```
Comment (connorlindsey) [inline public/app/plugins/datasource/tempo/datasource.ts:414]: We should only run these queries and display the table in the UI if the `tempoApmTable` toggle is enabled

Nit: Function names could be more descriptive (`firstServiceMapQuery` and `secondServiceMapQuery`)

**Review by connorlindsey — APPROVED** [review]
🏆 looks amazing!

File: docs/sources/datasources/tempo.md [hunk 3]
```diff
@@ -194,6 +194,25 @@ The color of each circle represents the percentage of requests in each of the fo
 
 Click on the service to see a context menu with additional links for quick navigation to other relevant information.
 
+## APM table
+
+The APM (Application Performance Management) table allows you to view several APM metrics out of the box.
```
Comment (brendamuir) [inline docs/sources/datasources/tempo.md:199]:
```suggestion
The APM (Application Performance Management) table enables you to view several APM metrics out of the box.
```

File: docs/sources/datasources/tempo.md [hunk 4]
```diff
@@ -194,6 +194,25 @@ The color of each circle represents the percentage of requests in each of the fo
...
+
+To display the APM table:
+
+- Enable the tempoApmTable feature flag in your ini file
+- Link a Prometheus datasource in the Tempo datasource settings
+- Navigate to [Explore]({{< relref "../explore/_index.md" >}})
+- Select the Tempo datasource
+- Select the **Service Graph** query type and run the query
+- Optionally, filter your results
```
Comment (brendamuir) [inline docs/sources/datasources/tempo.md:208]: Can we number these?

Comment (brendamuir) [inline docs/sources/datasources/tempo.md:208]:
```suggestion
1. Enable the tempoApmTable feature flag in your ini file
1. Link a Prometheus datasource in the Tempo datasource settings
1. Navigate to [Explore]({{< relref "../explore/_index.md" >}})
1. Select the Tempo datasource
1. Select the **Service Graph** query type and run the query
1. (Optional): filter your results
```

File: docs/sources/datasources/tempo.md [hunk 5]
```diff
@@ -194,6 +194,25 @@ The color of each circle represents the percentage of requests in each of the fo
...
+
+Note: The metric traces_spanmetrics_calls_total is used to display the name, rate & error rate columns and traces_spanmetrics_duration_seconds_bucket is used to display the duration column (these metrics will need to exist in your Prometheus datasource).
+
+Clicking on any row in the rate, error rate or duration columns will open up a query in Prometheus with that rows span name automatically set in the query. Cicking on a row in the links column will open up a query in Tempo with that rows span name automatically set in the query.
```
Comment (brendamuir) [inline docs/sources/datasources/tempo.md:212]:
```suggestion
Click a row in the rate, error rate, or duration columns to open a query in Prometheus with the span name of that row automatically set in the query. Click a row in the links column to open a query in Tempo with the span name of that row automatically set in the query.
```

**Comment by kvrhdn** [comment]
This is really cool! Thank you ❤️

Some feedback:

- in the table, there is a + and - 🔎 next to the span name but clicking on this doesn't do anything
- if you click on the rate it opens a sidepanel with this query:
  ```
  topk(5, sum(rate(traces_spanmetrics_calls_total{span_name="<span name>"}[$__rate_interval])) by (span_name))
  ```
  you can drop `topk(5, ...)`. Since we already filter on the span name, there will never be multiple results. So it becomes just:
  ```
  sum(rate(traces_spanmetrics_calls_total{span_name="<span name>"}[$__rate_interval])) by (span_name)
  ```
- Same as ⬆️ but for error rate
- What does the value in the rate column represent? Is that the average requests per second across the selected time range? Can we clarify this somehow, maybe with a tooltip or by naming the column something like `Rate (req/s)`?
- I'm not able to recreate the value shown in the rate column, the shown value seems to be almost double what I would expect (see screenshot). Besides using `$__range` and `@ end()`, is the query used by the table different?
  Weirdly enough, I don't have this issue with the error rate column. Values align with my direct promql query.
  Screenshot 2022-06-03 at 15 03 43
- If you add filters, the table is filtered down but these filters aren't passed on to the queries that open in a sidepanel.
  Screenshot 2022-06-03 at 15 13 27

Feedback/ideas for maybe a later PR:

- Something I just thought about: when displaying the error rate, it might be nicer to show both successful and failed requests. This shows the error rate in context. We can do this by keeping `span_status` out of the filter labels and sum by it:
  ```
  sum(rate(traces_spanmetrics_calls_total{span_name="<span name>"}[$__rate_interval])) by (span_name, span_status)
  ``` 
- it would be neat if you could add a filter from the context menu of a node in the service graph. Kind of like the + and - 🔎: there is a quick button to add/remove services from your filters.
- maybe add an option to configure the length of the table, 5 items is a good snapshot but it would be neat if you could increase it to 10 to get some more contex
- sorting doesn't query the underlying again, so if you sort by error rate you will not see the spans with the most errors overall but only of the 5 already shown.

**Comment by joey-grafana** [comment]
Thanks for the review!

- Removed + / - 
- I'm using the top(5 in the query as it returns the data sorted which is what we need for the table to show rows in order :) 
- `topk(5, sum(rate(traces_spanmetrics_calls_total{}[$__range] @ end())) by (span_name))` is what we use for displaying the rate col in the table. `topk(5, sum(rate(traces_spanmetrics_calls_total{span_name="HTTP Client"}[$__rate_interval])) by (span_name))` is what we use when clicking a link in the rate column.

with the above in mind, would you still like me to update the rate col name? Also, should I update the error rate col name too? 

perhaps you could Slack me your docker-compose file so I can see where you are generating metrics from (might tell why you were seing a double value)
- filters re-added

File: docs/sources/datasources/tempo.md [hunk 6]
```diff
@@ -190,6 +190,25 @@ The color of each circle represents the percentage of requests in each of the fo
 
 Click on the service to see a context menu with additional links for quick navigation to other relevant information.
 
+## APM table
+
+The APM (Application Performance Management) table enables you to view several APM metrics out of the box.
+
+To display the APM table:
+
+1. Enable the tempoApmTable feature flag in your ini file
```
Comment (brendamuir) [inline docs/sources/datasources/tempo.md:199]: Activate?

File: docs/sources/datasources/tempo.md [hunk 7]
```diff
@@ -190,6 +190,25 @@ The color of each circle represents the percentage of requests in each of the fo
 
 Click on the service to see a context menu with additional links for quick navigation to other relevant information.
 
+## APM table
+
+The APM (Application Performance Management) table enables you to view several APM metrics out of the box.
```
Comment (brendamuir) [inline docs/sources/datasources/tempo.md:195]:
```suggestion
The APM (Application Performance Management) table allows you to view several APM metrics out of the box.
```

File: docs/sources/datasources/tempo.md [hunk 8]
```diff
@@ -190,6 +190,25 @@ The color of each circle represents the percentage of requests in each of the fo
...
+
+To display the APM table:
+
+1. Enable the tempoApmTable feature flag in your ini file
+1. Link a Prometheus datasource in the Tempo datasource settings
+1. Navigate to [Explore]({{< relref "../explore/_index.md" >}})
+1. Select the Tempo datasource
+1. Select the **Service Graph** query type and run the query
+1. (Optional): filter your results
```
Comment (brendamuir) [inline docs/sources/datasources/tempo.md:204]: These should all end in a period

**Review by brendamuir — APPROVED** [review]

**Review by yvrhdn — APPROVED** [review]
Looks good to me! 🚀

**Comment by kvrhdn** [comment]
I have some unfortunate news though: we discovered that the metrics emitted by Tempo have a slightly different name from those from the agent and the OTel Collector. Our original intention was to align metric names and labels so it would be easy to use both, but somehow I still managed to use different names in Tempo 🙃 

We will be correcting this in Tempo (breaking change): https://github.com/grafana/tempo/pull/1478
The following metrics will change:
- `traces_spanmetrics_duration_seconds_bucket` -> `traces_spanmetrics_latency_bucket`
- `traces_spanmetrics_duration_seconds_count` -> `traces_spanmetrics_latency_count`
- `traces_spanmetrics_duration_seconds_sum` -> `traces_spanmetrics_latency_sum`

And the `span_status` label will become `status_code`.

I think the current PR uses the old metric names, we should switch this by the time this feature is released (does not have to be part of this PR).

**Pull request merged**